library/item/item_setting/src/main/java/org/quiltmc/qsl/item/setting/impl/RecipeRemainderLogicHandlerImpl.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.item.setting.impl;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.item.ItemStack;
+import net.minecraft.recipe.Recipe;
+import net.minecraft.screen.slot.Slot;
+import net.minecraft.util.ItemScatterer;
+import net.minecraft.util.collection.DefaultedList;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+import org.quiltmc.qsl.item.setting.api.RecipeRemainderLogicHandler;
+
+@ApiStatus.Internal
+public final class RecipeRemainderLogicHandlerImpl implements RecipeRemainderLogicHandler {
+	/**
+	 * @return {@code true} if returning the item was successful, or {@code false} if additional handling for the remainder is needed
+	 */
+	@Contract(mutates = "param1, param2")
+	private static boolean tryReturnItemToInventory(ItemStack remainder, DefaultedList<ItemStack> inventory, int index) {
+		ItemStack leftovers = inventory.get(index);
+		if (leftovers.isEmpty()) {
+			inventory.set(index, remainder);
+			return true;
+		}
+
+		return tryMergeStacks(leftovers, remainder);
+	}
+
+	/**
+	 * @return {@code true} if returning the item was successful, or {@code false} if additional handling for the remainder is needed
+	 */
+	@Contract(mutates = "param1, param2")
+	private static boolean tryReturnItemToSlot(ItemStack remainder, Slot slot) {
+		ItemStack leftovers = slot.getStack();
+		if (leftovers.isEmpty()) {
+			slot.setStack(remainder);
+			return true;
+		}
+
+		return tryMergeStacks(leftovers, remainder);
+	}
+
+	/**
+	 * @return {@code true} if the remainder stack was fully merged into the base stack, or {@code false}
```

Comment:
I was gonna ask for params too, but this is an internal class, it might not be necessary
```suggestion
	 * {@return {@code true} if the remainder stack was fully merged into the base stack, or {@code false} otherwise}
```